### PR TITLE
[Ready for review] Revamps bots, and adds the Bot call ability!

### PR DESCRIPTION
**Pull request by GunHog**
Poll justifying the continued effort for this project found here:
http://tgstation13.org/phpBB/viewtopic.php?f=15&t=418

Grants the AI the ability to call bots!

How it works:
- The AI uses "Access Bot Control" verb to bring up a popup containing all of the bots within the AI's Z level.
- From here, the AI can learn its general location and what it is currently doing.
- The AI can remotely access a bot's interface using the "Interface" button.
- The AI selects "Call", which puts it in "Waypoint Mode".
- The AI then clicks a valid, unobstructed spot on the station.
- The designated bot then responds to the call and attempts to move to chosen waypoint.
- While the bot is responding to the AI's call, it is granted all-access to prevent the AI having to pre-open doors to clear a path.
- If the bot happens to be off, it will be turned on automatically. (if the bot is not EMP'ed or something)
- Hacked bots gain a warning symbol before the name.
- <B>Manually emagged bots do not show up on the panel at all.</B>
- If a bot is unlocked, it can be removed from the AI's panel with a toggle. Only non-silicon users will see the toggle. Of course, this toggle will not function if the bot is hacked or emagged.

Patrolling enabled for floorbot and medibot! Both are now able to patrol the station, search for patients and hull breaches! Patrolling defaults to off for these types.

 _Bot Summoning!_
- All bots can be controlled remotely by an appropriate PDA cartridge!
- Engineers and atmos techs get floorbots.
- Medical staff get medibots.
- The Janitor gets the cleanbots.
- ED-209s are now included in the Securitron list. 

_Bot changes_
- Bots now use the nicer dark UI!
- Return of bridge mode for floorbots!
- <s>Medibots now dose inaprovaline in response to oxyloss instead of tricord.</s>

Current bot panel - 
screenshot 2014-06-26 10 18 10

New bot UI -
screenshot 2014-07-18 19 36 28

<B>Remie Edit: Approved for use of Topic Procs</B>

### CODE REVIEW

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 1]
```diff
@@ -30,6 +30,7 @@ var/list/ai_list = list()
 	var/obj/item/device/pda/ai/aiPDA = null
 	var/obj/item/device/multitool/aiMulti = null
 	var/obj/item/device/camera/ai_camera/aicamera = null
+	var/obj/machinery/bot/B = null
```
Comment (Giacom) [inline code/modules/mob/living/silicon/ai/ai.dm]: You shouldn't name variables like this when they aren't temporary.

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 2]
```diff
@@ -413,6 +417,20 @@ var/list/ai_list = list()
 			A.ai_actual_track(target)
 		return
 
+	if (href_list["callbot"]) //Command a bot to move to a selected location.
+		src.B = locate(href_list["callbot"]) in machines
```
Comment (Giacom) [inline code/modules/mob/living/silicon/ai/ai.dm]: This variable can be null.

Comment (Giacom) [inline code/modules/mob/living/silicon/ai/ai.dm]: Specifically, locate() can return null if it doesn't locate a value.

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 3]
```diff
@@ -509,6 +526,64 @@ var/list/ai_list = list()
...
+		src << "Wireless communication is disabled."
+		return
+
+	var/turf/ai_current_turf = get_turf(src)
+	var/ai_Zlevel = ai_current_turf.z
+	var/d
+	var/area/bot_area
+	d += "<table width='100%'><tr><td width='40%'><h3>Name</h3></td><td width='30%'><h3>Status</h3></td><td width='30%'><h3>Location</h3></td><td width='10%'><h3>Control</h3></td></tr>"
+
+	for (B in machines)
```
Comment (Giacom) [inline code/modules/mob/living/silicon/ai/ai.dm]: Use a local variable. I see no reason to keep it in a non-temporary variable.

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 4]
```diff
@@ -509,6 +526,64 @@ var/list/ai_list = list()
...
+	var/area/end_area = get_area(end_loc)
+	var/turf/start_loc = get_turf(B) //Get the bot's location.
+
+	//For giving the bot all-access.
+	var/obj/item/weapon/card/id/all_access = new /obj/item/weapon/card/id
+	var/datum/job/captain/All = new/datum/job/captain
+	all_access.access = All.get_access()
+
+	var/list/call_path = list()
+	call_path = AStar(start_loc, end_loc, /turf/proc/CardinalTurfsWithAccess, /turf/proc/Distance_cardinal, 0, 255, id=all_access)
```
Comment (Giacom) [inline code/modules/mob/living/silicon/ai/ai.dm]: If 255 is the maximum path finding range, I'd like it to be lowered to 128.

**Comment by Giacom** [comment]
There's a lot of code which should be in its own procs.

**Comment by Giacom** [comment]
I have only commented on a few specific cases but I see the same issues elsewhere, go through your code and look out for them.

**Comment by GunHog** [comment]
Thanks for the feedback Gia, it helps :)

File: code/game/machinery/bots/bots.dm [hunk 5]
```diff
@@ -14,8 +14,30 @@
...
+	#define BOT_ARREST			3	// arresting target
+	#define BOT_START_PATROL	4	// start patrol
+	#define BOT_PATROL			5	// patrolling
+	#define BOT_SUMMON			6	// summoned by PDA
+	#define BOT_CLEANING 		7	// cleaning (cleanbots)
+	#define BOT_REPAIRING		8	// repairing hull breaches (floorbots)
+	#define BOT_WORKING			9	// for MULEbots, when moving.
+	#define BOT_HEALING			10	// healing people (medbots)
+	#define BOT_RESPONDING		11	// responding to a call from the AI
+	var/list/busy_name = list("In Pursuit","Arresting","Arresting","Beginning Patrol","Patrolling","Summoned by PDA", \
```
Comment (RemieRichards) [inline code/game/machinery/bots/bots.dm]: > "Arresting","Arresting"

I'm aware its in the old list that was removed, but is it required?

Comment (GunHog) [inline code/game/machinery/bots/bots.dm]: It is used in the Beepsky and ED-209 process procs to determine if it is arresting or starting to arrest. The time difference is small enough to not really need its own unique text, but I can give it something to mark the difference. It is really there so I can keep the index lined up properly.

**Comment by GunHog** [comment]
Made some more tweaks and changes. On the note of making var/obj/machinery/bot/Bot a local variable, I find that it is used by several related procs, as well as the on-click. It would be redundant to declare it for every instance. I am still hunting for other improvements based on previous comments, however.

**Comment by ghost** [comment]
What's an easy way to test other people's PRs :s

**Comment by RemieRichards** [comment]
@paprka merge the code to a local branch. Play.

Alternatively, Learn to read code so you can spot areas that are bad, but hidden well.

File: code/game/machinery/bots/bots.dm [hunk 6]
```diff
@@ -14,7 +15,54 @@
...
+
+	var/blockcount = 0		//number of times retried a blocked path
+	var/awaiting_beacon	= 0	// count of pticks awaiting a beacon response
+
+	var/nearest_beacon			// the nearest beacon's tag
+	var/turf/nearest_beacon_loc	// the nearest beacon's location
+
+	var/beacon_freq = 1445		// navigation beacon frequency
+	var/control_freq = 1447		// bot control frequency
+	var/bot_type = "bot" //The type of bot it is, for radio control.
```
Comment (RemieRichards) [inline code/game/machinery/bots/bots.dm]: Don't use strings for this, use Constants or Defines (You used defines for Mode so Defines would probably be best for consistency) 

Better yet, replace bot_type with istype()

Comment (GunHog) [inline code/game/machinery/bots/bots.dm]: "bot_type" is used for "send_status()" which requires a string. It is an input for a radio signal to interact with beacons and PDAs. (see in the PDA folder "radio.dm" and "cart.dm") In this code, it is currently only used for security robots ("secbot") type and MULEs ("mulebot") types, as there are only two signal filters for bots. The other filters will be added to this PR once I am done coding them.

Comment (GunHog) [inline code/game/machinery/bots/bots.dm]: I have taken your advice and managed to make them into numerical defines. I would assume that performance wise, number defines would be faster than strings or istype(). Radio code procs very many times per second, so it might be the best option.

I will have it up for review in my next update.

Comment (RemieRichards) [inline code/game/machinery/bots/bots.dm]: that's good, and there will probably be a minor performance gain, only really in the fact that a 1 character number is less than a potentially 5+ character string.

**Comment by GunHog** [comment]
Merge conflict resolution is the scariest thing I have ever done on git. The meat on this beast is mostly finished. I just need to do some testing on this to find any potential bugs, then it should be good to go!

**Comment by GunHog** [comment]
This should be fully workable and test-ready!

**Comment by RemieRichards** [comment]
Well think of it this way @GunHog Who's generally considered our best tester?

I'll give you a hint, the name they go by starts with a G and ends in an unHog.

**Comment by GunHog** [comment]
RR, I have been testing this <b>Every day and after every update</b>. I would not have marked this for review if I were still finding more bugs. More testers is always better. Different minds give a different approach. The volunteers I have had so far have been amazing in helping me hunt things down.

**Comment by ike709** [comment]
pls merge must haz

**Comment by GunHog** [comment]
I will host a testing session of this tomorrow, and if we cannot find any more bugs, then I will mark this for merging / review. You may also download the code here: https://github.com/GunHog/NTstation13/archive/Bot_Call.zip

**Comment by ghost** [comment]
Tested each of the bots. Seems to be working for the most part, but I only tested calling them to my AI camera's location and shit. This is excellent gun hog. Maybe rogue AIs won't need to rely on sec borgs as much now that they can weaponize securitrons more effectively.

**Comment by GunHog** [comment]
Thanks for your contribution! I am giving this another once over tomorrow, then I will flag it for merging.

As for using securitrons as weapons, it is certainly viable, but dangerous. They are quite fragile, so keeping the element of surprise is critical in keeping them alive.

**Comment by GunHog** [comment]
Conflicts resolved, but this is currently untested in its latest form, as well as having an exploit fix possibly lost. I may need to modify A-star to ensure that exploit is fixed. There are some additional tweaks and tests I need to do as well, especially to ensure any content I modified in the conflict fix is still functional.

**Comment by GunHog** [comment]
Summary of latest update: <S> Medibots use inaproavline to respond to oxyloss now</S>, I mirrored the attack cooldown for bots on /tg/, gave the dark UI to bots (Special thanks to Ikarrus), and centralized the Bot topics.

I have begun a new round of testing due to the merge conflict fix and logical changes to A-star. One day, just one day, this will actually be merge ready.

**Comment by GunHog** [comment]
I guess this is as ready as it is ever going to be. Any further bugs would probably be found in live play. Ready when you guys complete your review, @NTStation13Admin

**Comment by GunHog** [comment]
One last minute change!
- Emagging or hacking a bot locks the interface, preventing access to the settings panel.
- Humans cannot interact with hacked or emagged bots, period.
- Silicons can control hacked bots, but cannot interact with manually emagged bots.

*A minor bug exists where a Silicon can change a bot's type specific settings even if it is manually emagged. It is not considered an issue, as emagged bots do not respect their settings anyway.

**Comment by GunHog** [comment]
@RemieRichards I assume you will be my reviewer for today?

**Comment by RemieRichards** [comment]
I'm busy today @GunHog I'll look into it tomorrow (Tuesday)

**Comment by RemieRichards** [comment]
@GunHog 

Right so the only thing I was worried about with this really was the use of the A\* Algorithm and that's all fine.

You told me you didn't notice any bugs. (besides the one we talked about, which is an edge case and can be fixed later on) so as long as that's still true, give me an @RemieRichards and I'll merge.

**Comment by GunHog** [comment]
Wonderful! Please give me a bit to run a final test on everything, then I will ping you! :D

**Comment by GunHog** [comment]
Sorry for the supermassive wait, @RemieRichards . Had some last minute things to fix found in the last multiplayer playtest! Should be ready now! :D

**Comment by RemieRichards** [comment]
Alrighty then!
